util/udma_barrier.h

```diff
@@ -198,12 +198,21 @@
 #define mmio_flush_writes() asm volatile("membar #StoreStore" ::: "memory")
 #elif defined(__aarch64__)
 #define mmio_flush_writes() asm volatile("dsb st" ::: "memory");
-#elif defined(__sparc__) || defined(__s390x__)
+#elif defined(__sparc__)
 #define mmio_flush_writes() asm volatile("" ::: "memory")
 #elif defined(__loongarch__)
 #define mmio_flush_writes() asm volatile("dbar 0" ::: "memory")
 #elif defined(__riscv)
 #define mmio_flush_writes() asm volatile("fence ow,ow" ::: "memory")
+#elif defined(__s390x__)
+#include "s390_mmio_insn.h"
```

Comment:
Why is read and write being defined through udma_barrier.h ?

Reply:
I include the `s390_mmio_insn.h` for the `s390_is_mio_supported` flag. One way not to include the read/write would be to move the `MAKE_READ/MAKE_WRITE` to `mmio.h` and keep `s390_mmio_insn.h` just for the flag and the instruction wrappers. Would that be acceptable @jgunthorpe?

Reply:
You can just add an extern for the flag, it doesn't have be in exactly one header. Just be sure to include all the headers that define the extern in the C file that instantiates it.

Reply:
@jgunthorpe I tried that, it causes a redundant redecleration warning in all files which include both.

Reply:
I tried that, sadly it causes a redundant redecleration warning in all files that include both `udma_barrier.h` and `mmio.h`.

Reply:
Now clearly we could work around this with an extra define much like a header guard such that `util/udma_barrier.h` or `util/s390_mmio_insn.h` depending on which comes first. But that's kind of ugly. No?

Personally I think, moving the MAKE_READ/MAKE_WRITE as well as the 8 bit read/write to `util/mmio.h` while including `util/s390_mmio_insn.h` is a clean solution. It even improves the match with the file name, keeping `s390_mmio_insn.h` strictly about the instructions. I also noticed I had the flag check in both the `s390_pciwb()` itself and again in `mmio_flush_writes()` with that removed I find it pretty clean. Basically in `util/udma_barrier.h` we're just left with `#define mmio_flush_writes() s390_pciwb()`

Reply:
I've pushed a v6 with the above idea for you to take a look.

Reply:
Oh blah, yes we have to deal with that extra warning in these headers.